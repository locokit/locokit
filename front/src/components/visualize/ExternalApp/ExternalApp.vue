<template>
  <iframe v-if="src" :src="src" />
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { LckTableRow } from '@/services/lck-api/definitions'
import { ExternalAppURLPart, EXTERNAL_APP_URL_PART_TYPE } from '@locokit/lck-glossary/src'

export default Vue.extend({
  name: 'ExternalApp',
  props: {
    content: {
      type: Object as PropType<Record<string, LckTableRow | null>>,
    },
    settings: {
      type: Object as PropType<{ parts: ExternalAppURLPart[]}>,
    },
  },
  computed: {
    src () {
      return (this.settings as { parts: ExternalAppURLPart[]})?.parts.map(currentPart => {
        switch (currentPart.type) {
          case EXTERNAL_APP_URL_PART_TYPE.STRING:
            return currentPart.string
          case EXTERNAL_APP_URL_PART_TYPE.SOURCE:
            const currentSource = (((this.content) as Record<string, LckTableRow>)[currentPart.id as string] as LckTableRow)
            if (!currentSource) return null // then the join will be null, and we don't display the iframe before the url is complete
            if (currentPart.fieldId === 'id') {
              return currentSource.id
            } else {
              return currentSource.data[currentPart.fieldId as string]
            }
        }
      }).join('')
    },
  },
})
</script>

<style scoped>
iframe {
  border: unset;
  width: 100%;
  min-height: 500px;
  height: 100%;
}
</style>
