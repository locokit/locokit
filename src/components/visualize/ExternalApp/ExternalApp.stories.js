import ExternalApp from './ExternalApp'

export default {
  title: 'components/visualize/ExternalApp',
  component: ExternalApp,
}
export const defaultStory = () => (
  {
    components: { ExternalApp },
    template: '<ExternalApp />',
  }
)

defaultStory.storyName = 'default'

const blockExternalApp = {
  id: 1,
  title: 'My ExternalApp\'s block',
  type: 'ExternalApp',
  settings: {
    parts: [{
      type: 'String',
      string: 'https://',
    }, {
      type: 'Source',
      id: 'first-source',
      fieldId: 'id',
    }, {
      type: 'Source',
      id: 'second-source',
      fieldId: '72205f2e-a9cc-4e30-82ba-6f1f6e357e0c',
    }, {
      type: 'String',
      string: '.io',
    }],
  },
  content: {
    'first-source': {
      id: 'demo.',
    },
    'second-source': {
      data: {
        '72205f2e-a9cc-4e30-82ba-6f1f6e357e0c': 'locokit',
      },
    },
  },
}

export const withParts = (args, { argTypes }) => (
  {
    components: { ExternalApp },
    props: Object.keys(argTypes),
    template: '<ExternalApp v-bind="block" />',
  }
)

withParts.storyName = 'with parts in settings'
withParts.args = {
  block: blockExternalApp,
}
