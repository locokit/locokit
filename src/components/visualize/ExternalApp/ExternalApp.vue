<template>
  <iframe
    :src="src"
  />
</template>

<script>
export default {
  name: 'ExternalApp',
  props: {
    src: {
      type: String,
      required: true,
    },
  },
}
</script>

<style scoped>
iframe {
  border: unset;
  width: 100%;
  height: 100%;
}
</style>
