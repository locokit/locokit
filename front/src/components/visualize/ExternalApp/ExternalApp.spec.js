import { shallowMount } from '@/../tests/unit/local-test-utils'
import ExternalApp from '@/components/visualize/ExternalApp/ExternalApp.vue'

const blockExternalApp = {
  id: 1,
  title: 'My TableSet\'s block',
  type: 'TableSet',
  settings: {
    parts: [{
      type: 'String',
      string: 'https://',
    }, {
      type: 'Source',
      id: 'first-source',
      fieldId: 'id',
    }, {
      type: 'Source',
      id: 'second-source',
      fieldId: '72205f2e-a9cc-4e30-82ba-6f1f6e357e0c',
    }, {
      type: 'String',
      string: '.io',
    }],
  },
  content: {
    'first-source': {
      id: 'docs.',
    },
    'second-source': {
      data: {
        '72205f2e-a9cc-4e30-82ba-6f1f6e357e0c': 'locokit',
      },
    },
  },
}

describe('ExternalApp', () => {
  it('match snapshot without any settings', async () => {
    const wrapper = await shallowMount(ExternalApp)
    expect.assertions(1)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('display the iframe to docs locokit', async () => {
    const wrapper = await shallowMount(ExternalApp, {
      propsData: {
        settings: blockExternalApp.settings,
        content: blockExternalApp.content,
      },
    })
    expect.assertions(3)
    const iframe = wrapper.findAll('iframe')
    expect(iframe.length).toBe(1)
    expect(iframe.at(0).attributes().src).toBe('https://docs.locokit.io')
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('display an iframe with broken URL if one of the part is null', async () => {
    const wrapper = await shallowMount(ExternalApp, {
      propsData: {
        settings: blockExternalApp.settings,
        content: {
          ...blockExternalApp.content,
          'second-source': null,
        },
      },
    })
    expect.assertions(3)
    const iframe = wrapper.findAll('iframe')
    expect(iframe.length).toBe(1)
    expect(iframe.at(0).attributes().src).toBe('https://docs..io')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
