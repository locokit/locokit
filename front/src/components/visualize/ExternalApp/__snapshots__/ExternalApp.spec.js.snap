// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalApp display an iframe with broken URL if one of the part is null 1`] = `<iframe src="https://docs..io"></iframe>`;

exports[`ExternalApp display the iframe to docs locokit 1`] = `<iframe src="https://docs.locokit.io"></iframe>`;

exports[`ExternalApp match snapshot without any settings 1`] = `""`;
